librad/src/net/discovery/mdns.rs

```diff
@@ -0,0 +1,230 @@
+                        ResourceRecord::IN(&cname, RData::a(*ip4.ip()))
+                            .set_ttl(Duration::from_secs(300)),
+                    );
+                },
+                V6(ip6) => {
+                    packet.add_answer(
+                        ResourceRecord::IN(&cname, RData::aaaa(*ip6.ip()))
+                            .set_ttl(Duration::from_secs(300)),
+                    );
+                },
+            }
+        }
+        packet.header_mut().set_id(rand::random()).set_query(false);
+        let packet = packet.build();
+        srv.enqueue_response(packet);
+    }
+}
+
+fn discover(cname: &str, resp: mdns::Response) -> impl Iterator<Item = (PeerId, Vec<SocketAddr>)> {
+    use mdns::{Record, RecordKind::*};
+
+    let mut ports = BTreeMap::<PeerId, BTreeSet<u16>>::new();
+    let mut addrs = BTreeMap::<PeerId, BTreeSet<IpAddr>>::new();
+    let mut add_addr = |peer, addr| {
+        addrs
+            .entry(peer)
+            .and_modify(|addrs| {
+                addrs.insert(addr);
+            })
+            .or_insert_with(|| vec![addr].into_iter().collect());
+    };
+
+    fn peer_from_domain(dom: &str) -> Option<PeerId> {
+        dom.split('.').next().and_then(|first| first.parse().ok())
+    }
+
+    for Record { name, kind, .. } in resp.records() {
+        match kind {
+            SRV { target, port, .. } if target != cname && name.ends_with(SERVICE_NAME) => {
+                if let Some(peer) = peer_from_domain(target) {
+                    ports
+                        .entry(peer)
+                        .and_modify(|ports| {
+                            ports.insert(*port);
+                        })
+                        .or_insert_with(|| vec![*port].into_iter().collect());
+                }
+            },
+            A(ip4) if name != cname => {
+                if let Some(peer) = peer_from_domain(name) {
+                    add_addr(peer, IpAddr::from(*ip4));
+                }
+            },
+            AAAA(ip6) if name != cname => {
+                if let Some(peer) = peer_from_domain(name) {
+                    add_addr(peer, IpAddr::from(*ip6));
+                }
+            },
+
+            _ => (),
+        }
+    }
+
+    addrs.into_iter().filter_map(move |(peer, addrs)| {
+        ports.get(&peer).map(|ports| {
+            let addrs = addrs
+                .iter()
+                .flat_map(|addr| ports.iter().map(move |port| SocketAddr::new(*addr, *port)))
+                .collect();
+            (peer, addrs)
+        })
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use pretty_assertions::assert_eq;
+    use std::net::Ipv4Addr;
+
+    use crate::{keys::SecretKey, peer::PeerId};
+
+    // `madness` uses tokio, so no choice here -- MADNESS!
+    #[tokio::test]
+    // CI doesn't support multicast UDP
+    #[ignore]
+    async fn ohai() {
+        let lolek_id = PeerId::from(SecretKey::new());
```

Comment:
![image](https://user-images.githubusercontent.com/1585/100805786-9900b600-342f-11eb-8631-9cd04b1768fe.png)

Reply:
I **need** to know the name of the pirate